File: ios/Sources/RoadSenseNSBootstrap/Network/UploadEligibilityPolicy.swift
import Foundation

public enum NetworkPathStatus: Equatable, Sendable {
    case satisfied
    case unsatisfied
}

public struct NetworkPathSnapshot: Equatable, Sendable {
    public let status: NetworkPathStatus
    public let isExpensive: Bool

    public init(status: NetworkPathStatus, isExpensive: Bool) {
        self.status = status
        self.isExpensive = isExpensive
    }
}

public enum UploadEligibilityPolicy {
    public static func shouldUpload(
        pendingCount: Int,
        network: NetworkPathSnapshot,
        nextAttemptAt: Date? = nil,
        now: Date = Date()
    ) -> Bool {
        guard network.status == .satisfied else {
            return false
        }

        guard pendingCount > 0 else {
            return false
        }

        if let nextAttemptAt, nextAttemptAt > now {
            return false
        }

        return true
    }
}

File: android/core-api/src/main/kotlin/ca/roadsense/ns/api/UploadEligibilityPolicy.kt
package ca.roadsense.ns.api

import java.time.Instant

enum class NetworkPathStatus { SATISFIED, UNSATISFIED }

data class NetworkPathSnapshot(
    val status: NetworkPathStatus,
    val isExpensive: Boolean,
)

object UploadEligibilityPolicy {
    fun shouldUpload(
        pendingCount: Int,
        network: NetworkPathSnapshot,
        nextAttemptAt: Instant? = null,
        now: Instant = Instant.now(),
    ): Boolean {
        if (network.status != NetworkPathStatus.SATISFIED) return false
        if (pendingCount <= 0) return false
        if (nextAttemptAt != null && nextAttemptAt > now) return false
        return true
    }
}
